language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+  newparam(:user_data) do
+  end
+end
+```
+
+### newmetaparam
+
+Defines a new metaparameter for a type. Puppet has several built-in metaparameters described below. It is uncommon for types to define new metaparams as these generally require puppet core changes to be useful, e.g. `noop`, `tag`, `schedule`, etc.
+
+### validate
+
+Performs per-resource validation at catalog application time. This method is often used to validate related parameters and properties for a single resource, e.g. mutually exclusive properties or if one property requires another. One concrete example is the `file` resource requires the `source` parameter to be set if the `recurse` parameter is set to `remote`:
+
+```ruby
+Puppet::Type.newtype(:file) do
+  validate do
+    self.fail "You cannot specify a remote recursion without a source" if !self[:source] && self[:recurse] == :remote
+  end
+end
+```
+
+### autorelations
+
+Allows a type to define automatic relationships (`before`, `subscribe`, `require`, `notify`) between each instance of the type and instances of a different type. For example, the `exec` type will autorequire any file resource whose `path` matches the exec  `command`. Note the relationship will only be added to the catalog if puppet is managing both ends of the relationship. That way puppet will automatically create the file containing the command to execute, before trying to executing it, regardless of manifest order.
+
+```ruby
+autorequire(:file) do
+  reqs = []
+  self[:command].scan(file_regex) { |str|
+    reqs << str
+  }
+  reqs
+end
+```
+
+Autorequires are by far the most common, though recently puppet added support for `autobefore`, `autonotify`, and `autosubscribe`.
+
+### pre_run_check
+
+The agent will call the `pre_run_check` method for each `Puppet::Type` instance in the agent's catalog. This provides an opportunity for a resource to perform consistency checks/validation against other resources in the catalog. It differs from the `validate` method, since it is called later during catalog application, and can rely on the catalog having all generated resources.
+
+### provider features
+
+Defines a feature for the type, which allows puppet to perform additional validation on the agent at catalog application time based on the selected provider. For example the `service` type defines an `enableable` feature:
+
+```ruby
+Puppet::Type.newtype(:service) do
+  feature :enableable
+end
+```
+
+And the type specifies that the `enable` property can only be managed if the selected provider supports the `enableable` feature. This validation check is performed at catalog application time, once the provider has been resolved on the agent.
+
+```ruby
+Puppet::Type.newtype(:service) do
+  newproperty(:enable, :required_features => :enableable) do
+  end
+end
+```
+
+A **provider** indicates it supports the feature using `has_feature`:
+
+```ruby
+Puppet::Type.type(:service).provide :launchd, :parent => :base do
+  has_feature :enableable
```

Comment:
I don't think that will work, because the catalog is not in scope when evaluating provider features. A provider either statically declares it has that feature:

```ruby
has_feature :manages_symlinks
```

Or the provider feature can be conditional on some other global state like osversion, if dependent gem is present, etc:

```ruby
has_feature :libuser if Puppet.features.libuser?
has_feature :system_groups unless %w{HP-UX Solaris}.include? Facter.value(:operatingsystem)
```